agent/stats/engine.go

```diff
@@ -163,77 +199,78 @@ func (engine *DockerStatsEngine) removeAll() {
-	}
 
-	for _, containerID := range listContainersResponse.DockerIDs {
-		engine.addContainer(containerID)
+	for task := range engine.tasksToHealthCheckContainers {
+		delete(engine.tasksToContainers, task)
 	}
-
-	return nil
 }
 
-// addContainer adds a container to the map of containers being watched.
-// It also starts the periodic usage data collection for the container.
-func (engine *DockerStatsEngine) addContainer(dockerID string) {
-	engine.lock.Lock()
-	defer engine.lock.Unlock()
-
+// addContainerUnsafe adds a container to the map of containers being watched.
+func (engine *DockerStatsEngine) addContainerUnsafe(dockerID string) (*StatsContainer, error) {
 	// Make sure that this container belongs to a task and that the task
 	// is not terminal.
 	task, err := engine.resolver.ResolveTask(dockerID)
 	if err != nil {
-		seelog.Debugf("Could not map container to task, ignoring, err: %v, id: %s", err, dockerID)
-		return
+		return nil, errors.Wrapf(err, "could not map container to task, ignoring container: %s", dockerID)
 	}
 
 	if len(task.Arn) == 0 || len(task.Family) == 0 {
-		seelog.Debugf("Task has invalid fields, id: %s", dockerID)
-		return
+		return nil, errors.Errorf("stats add container: invalid task fields, arn: %s, familiy: %s", task.Arn, task.Family)
 	}
 
 	if task.GetKnownStatus().Terminal() {
-		seelog.Debugf("Task is terminal, ignoring, id: %s", dockerID)
-		return
+		return nil, errors.Errorf("stats add container: task is terminal, ignoring container: %s, task: %s", dockerID, task.Arn)
+	}
+
+	seelog.Debugf("Adding container to stats watch list, id: %s, task: %s", dockerID, task.Arn)
+	statsContainer := newStatsContainer(dockerID, engine.client, engine.resolver)
+	engine.tasksToDefinitions[task.Arn] = &taskDefinition{family: task.Family, version: task.Version}
+
+	if !engine.disableMetrics {
+		// Adding container to the map for collecting stats
+		statsContainer = engine.addToStatsContainerMapUnsafe(task.Arn, dockerID, statsContainer, engine.containerMetricsMapUnsafe)
+	}
+
+	dockerContainer, err := engine.resolver.ResolveContainer(dockerID)
+	if err == nil && dockerContainer.Container.HealthStatusShouldBeReported() {
+		seelog.Debugf("Adding container to stats health check watch list, id: %s, task: %s", dockerID, task.Arn)
+		engine.addToStatsContainerMapUnsafe(task.Arn, dockerID, statsContainer, engine.healthCheckContainerMetricsMapUnsafe)
 	}
 
+	return statsContainer, nil
+}
+
+func (engine *DockerStatsEngine) containerMetricsMapUnsafe() map[string]map[string]*StatsContainer {
+	return engine.tasksToContainers
+}
+func (engine *DockerStatsEngine) healthCheckContainerMetricsMapUnsafe() map[string]map[string]*StatsContainer {
+	return engine.tasksToHealthCheckContainers
+}
+
+func (engine *DockerStatsEngine) addToStatsContainerMapUnsafe(
+	taskARN, containerID string,
+	statsContainer *StatsContainer,
+	statsMapToUpdate func() map[string]map[string]*StatsContainer) *StatsContainer {
```

Comment:
why is this method returning `*StatsContainer`? iiuc, you're just returning `statsContainer` here and that's not changing any state. It may make more sense to return a bool indicating if we should start watching this container or not based on its existence in the map.

Reply:
I was trying to avoid an extra `if` check, but I can modify that if this makes the code easier to read.